### PR TITLE
Fix FillConsoleOutputCharacterA crash

### LINKED ISSUE #4258: FillConsoleOutputCharacterA crashes conhost when passed an invalid character
# Environment

```none
Windows build number: Version 10.0.18362.535
```

# Steps to reproduce

Compile and run the following C program in a conhost shell:

```c
#include <windows.h>

void main() {
    SetConsoleOutputCP(50220);
    HANDLE handle = GetStdHandle(STD_OUTPUT_HANDLE);
    DWORD written;
    FillConsoleOutputCharacterA(handle, 14, 1, COORD{ 0,0 }, &written); 
}
```
# Expected behavior

I believe codepoint 14 is invalid in the given codepage, so I would expect it to write out something like the unicode replacement character in the top left corner of the screen buffer, or possibly nothing at all. The legacy console seems to write out a null character.

# Actual behavior

The conhost crashes.

What's happening is that the `FillConsoleOutputCharacterAImpl` method is calling `ConvertToW` with a string that can't be converted in the given codepage. And `ConvertToW` then throws an exception when `MultiByteToWideChar` returns 0.

https://github.com/microsoft/terminal/blob/9e5792ba51236945b942e560469e1abd7fb93c22/src/types/convert.cpp#L41-L42

And note that `FillConsoleOutputCharacterAImpl` is declared as `noexcept`, even though it quite clearly is capable of throwing exceptions. I've actually noticed a few cases like that - we may need to do an audit of our `noexcept` usage, and make sure it's being applied appropriately.

### PR DESCRIPTION
**Pull request by mkitzan**
## Summary of the Pull Request
Despite being specified as `noexcept`, `FillConsoleOutputCharacterA` emits an exception when a call to `ConvetToW` is made with an argument character which can't be converted. This PR fixes this throw, by wrapping `ConvertToW` in a try-catch_return.

 
## PR Checklist
* [x] Closes #4258
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed: thanks @miniksa 
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place

## Detailed Description of the Pull Request / Additional comments
Following the symantics of other `FillConsoleOutputCharacter*` the output param `cellsModified` is set to `0`. The try-catch_return is also what other functions of this family perform in case of errors.

## Validation Steps Performed
~The repro test in the issue uses this function as it exists in the `windows.h` library not the one present in the WT repo (by my understanding). Because of this, I have not been able to test that this change corrects the original repro.~
Original repro no longer crashes.

### CODE REVIEW

**Comment by zadjii-msft** [comment]
> The repro test in the PR uses this function as it exists in the `windows.h` library not the one present in the WT repo (by my understanding). Because of this, I have not been able to test that this change corrects the original repro.

So, fortunately, you actually can test this. The functions that are in `windows.h` are basically just the client-side implementation of these API calls, that are _serviced_ by `conhost.exe`. So when you make a change to `conhost.exe` and run the `OpenConosle.exe` that's built from this repo, you can actually check if the change in implementation fixes the bug. You should be able to run the sample repro program in an `OpenConsole.exe` instance and ensure that it fixes the bug. 

<hr>

Are there any other callers of `ConvertToW` (outside of tests) that aren't try/catching the call as well?

**Comment by mkitzan** [comment]
@zadjii-msft thanks for the info on testing, and a good thing too.

Inside [`ConvertToW`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L29) there's something causing a crash not a throw. The line that's sticking out to me in `ConvertToW` is this call to [`MultiByteToWideChar`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L41) the destination buffer is a `nullptr` while the return value is still being used to determine the output `wstring` size later. The symantics of [`MultiByteToWideChar`](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar) don't seem to cover this usage. I'd love to hear a second opinion on this.

I haven't done a pass for other `ConvertToW` calls. I'll do one once the `ConvertToW` crash is figured out.

**Comment by mkitzan** [comment]
I'll take a closer look at the crash this weekend. School has been eating up all my time this week.

**Comment by mkitzan** [comment]
Here's the crash: `string_view.data()` is not inherently null terminated and we pass a [`string_view` of a single char](https://github.com/microsoft/terminal/blob/master/src/host/_output.cpp#L310) to [`ConvertToW`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L29) then `MultiByteToWideChar` uses the pointer from `data()` to walk the buffer resulting in a crash.

@zadjii-msft, with the latest commit the repro no longer crashes. I did a `noexcept` check for other `ConvertToW` callers and all other calls are already wrapped in try-catch blocks.

**Comment by miniksa** [comment]
> @zadjii-msft Mike Griese FTE thanks for the info on testing, and a good thing too.
> 
> Inside [`ConvertToW`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L29) [](/microsoft/terminal/blob/HEAD@{2020-01-21T18:46:44Z}/src/types/convert.cpp#L29) there's something causing a crash not a throw. The line that's sticking out to me in `ConvertToW` is this call to [`MultiByteToWideChar`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L41) [](/microsoft/terminal/blob/HEAD@{2020-01-21T18:46:44Z}/src/types/convert.cpp#L41) the destination buffer is a `nullptr` while the return value is still being used to determine the output `wstring` size later. The symantics of [`MultiByteToWideChar`](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar) don't seem to cover this usage. I'd love to hear a second opinion on this.

The important part is the final parameter, `cchWideChar` being 0. Setting that requests that MB2WC tell us how many wchar_ts it will need (including a null terminator) if it were to actually give us the result. This means that the destination buffer pointer `lpWideCharStr` doesn't matter because this call is only attempting to retrieve the count. 

The MB2WC function should happily work with both counted and null-terminated (z) strings.

> I haven't done a pass for other `ConvertToW` calls. I'll do one once the `ConvertToW` crash is figured out.


> Here's the crash: `string_view.data()` is not inherently null terminated and we pass a [`string_view` of a single char](https://github.com/microsoft/terminal/blob/master/src/host/_output.cpp#L310) [](/microsoft/terminal/blob/HEAD@{2020-01-24T05:13:31Z}/src/host/_output.cpp#L310) to [`ConvertToW`](https://github.com/microsoft/terminal/blob/master/src/types/convert.cpp#L29) [](/microsoft/terminal/blob/HEAD@{2020-01-24T05:13:31Z}/src/types/convert.cpp#L29) then `MultiByteToWideChar` uses the pointer from `data()` to walk the buffer resulting in a crash.

See, I don't know about this. I swear it won't walk past the count that it's given for the source parameter. It will if the count was `-1` per the documentation... but with an accurate count of `1`, it should have no problem.

> @zadjii-msft Mike Griese FTE, with the latest commit the repro no longer crashes. I did a `noexcept` check for other `ConvertToW` callers and all other calls are already wrapped in try-catch blocks.





I think adding the try-catch block here is definitely warranted. It would probably harm nothing to just function-level TRY the whole method instead of having to declare the string outside the try and then use fill it later.

Finally, I think we're going to need a test here that confirms the error code coming back from an invalid/unconvertible input character is the same error code that the v1 console gave. I'd want to see that in feature tests.

Given that this is a bit weird in that you are seeing a crash that I cannot believe exists, I'd like to try to repro this myself. If I can get some time...

**Comment by miniksa** [comment]
OK I adapted the original repro scenario from #4258 into a feature test.

I ran it against conhost v1 and this was the behavior:

- Set the codepage to 50220 with SetConsoleOutputCP.
- Fill with the invalid character (0x14) via FillConsoleOutputCharacterA --> function call succeeds, written = 1.
- Read back the same position via ReadConsoleOutputCharacterA --> function succeeds, read = 1, value of character read = 0x0.

I also reproduced the crash against v2. We should likely match the v1 behavior, beyond just fixing the crash, if it isn't matching. Will continue looking.

**Comment by mkitzan** [comment]
Interesting that there's this discrepancy between v1 and v2. I'll take some time this weekend to have another look. Thanks for the v1 repro scenario, @miniksa, I'll try and get the behavior to match it.

**Comment by miniksa** [comment]
@mkitzan, I'll keep looking now. And upload the test. We can merge our two together to fix it.

**Comment by miniksa** [comment]
https://github.com/microsoft/terminal/commit/f31a3c1f7448fcde2badd6b6efbe9c07dab713cc

**Comment by miniksa** [comment]
Part of the problem here also seems to be that `ConvertToW:41` seems to return 0, which is supposed to be a failure code here: https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar.

So then we call `THROW_LAST_ERROR_IF` but that itself throws because the last error is set to 0 which is the "OK" error code.

**Comment by miniksa** [comment]
I'm going to try to debug the v1 host to see how it is coming to a different translation conclusion than the v2 one.

**Comment by miniksa** [comment]
OK. So the v1 host was awful here.

It would do...
```
WCHAR ConversionFunction(UINT codepage, char ch)
{
	WCHAR wch = 0;
	MultiByteToWideChar(codepage, 0, &ch, 1, &wch, 1);
	return wch;
}
```

That's right. It didn't check if there was an error. It just tried it and then plowed ahead with the 0 value if it didn't work. 

That feels terribly bad.

I think what we do here is update the test I just wrote to document the difference here, fix this so it doesn't crash (and returns an acceptable error), and then check it in and roll from there.

File: src/host/ft_host/API_FillOutputTests.cpp [hunk 1]
```diff
@@ -10,6 +10,24 @@ class FillOutputTests
     BEGIN_TEST_CLASS(FillOutputTests)
     END_TEST_CLASS()
 
+    // Adapted from repro in GH#4258
+    TEST_METHOD(FillWithInvalidCharacterA)
+    {
+        VERIFY_WIN32_BOOL_SUCCEEDED(SetConsoleOutputCP(50220));
```
Comment (DHowett-MSFT) [inline src/host/ft_host/API_FillOutputTests.cpp]: this is run in isolation, yes?

Comment (miniksa) [reply src/host/ft_host/API_FillOutputTests.cpp]: No but it should be. Good catch.

**Comment by mkitzan** [comment]
> That's right. It didn't check if there was an error. It just tried it and then plowed ahead with the 0 value if it didn't work.

Well that's something, eh? Sorry for being AWOL on this one recently: it's midterm season.

**Comment by miniksa** [comment]
> > That's right. It didn't check if there was an error. It just tried it and then plowed ahead with the 0 value if it didn't work.
> 
> Well that's something, eh? Sorry for being AWOL on this one recently: it's midterm season.

Not a problem. As long as you don't mind me partying on your branch.

**Comment by mkitzan** [comment]
> Not a problem. As long as you don't mind me partying on your branch.

Always up for a party, thanks for the help!

**Comment by miniksa** [comment]
OK I cannot comprehend why the test got back a 0x2 error code instead of the 0xffff I was expecting, but then I went to go figure out why all tests are being bad right now in #4490. So I'll get back to this.

**Comment by DHowett-MSFT** [comment]
:tada: Once again, thanks for the contribution!

This pull request was included in a set of conhost changes that was just
released with [Windows Insider Build 19603](https://blogs.windows.com/windowsexperience/2020/04/08/announcing-windows-10-insider-preview-build-19603/).

**Pull request merged**